201-api-management-logs-to-moesif-using-eventhub-webapp/README.md

```diff
@@ -0,0 +1,129 @@
+
+          var headers = context.Response.Headers
+                                          .Select(h => string.Format("{0}: {1}", h.Key, String.Join(", ", h.Value)))
+                                          .ToArray<string>();
+
+          var headerString = (headers.Any()) ? string.Join("\r\n", headers) + "\r\n" : string.Empty;
+
+          return "response:"  + context.Variables["message-id"] + "\n"
+                              + statusLine + headerString + "\r\n" + body;
+     }
+  </log-to-eventhub>
+  </outbound>
+</policies>
+```
+The [ARM Template `nested/microsoft.apimanagement/service/apis/policies.json`](nested/microsoft.apimanagement/service/apis/policies.json) can be used as a reference to set policy XML thru Azure deployment. This sample template, as implemented, will overwrite existing policy.xml
+That's it! API logs should start showing up after a few minutes.
+
+## Manual creation
+
+If you want to create the resources directly, follow [Microsoft's documentation](https://docs.microsoft.com/en-us/azure/api-management/api-management-log-to-eventhub-sample) on configuring Moesif. The example WebJob is available [on GitHub](https://github.com/Moesif/ApimEventProcessor).
+
+## About this template:
+This template performs the following tasks
+
+- Create Azure Eventhub Namespace, an Eventhub and a policy for Send and for Listen
+- If an existing Azure Api Management instance is provided, create a new logger `log-to-eventhub` named `moesif-log-to-event-hub` for use in Api Management policy xml.
+- Create Storage account to store eventhub consumption checkpoints
+- Create Azure App Service plan and an Azure App Service. Sets environment variables for use by Webjob
+- Create a Webjob (Continuous) that listens for eventhub events (Api Management request/responses) and forwards events to Moesif.
+
+- The Webjob pulls [Moesif/ApimEventProcessor](https://github.com/Moesif/ApimEventProcessor), builds it, and runs it using preset environment variable parameters. (At present, this webjob is not automatically run. It can be manually launched by deploying arm template [nested/microsoft.web/sites/extensions.json](nested/microsoft.web/sites/extensions.json) and providing it path to zip file [scripts/apim-2-moesif-webjob-webdeploy.zip](scripts/apim-2-moesif-webjob-webdeploy.zip))
+
+If an existing Azure Api Management is not specified, the `log-to-eventhub` logger is not automatically created. To manually create this logger, utilize the [`nested/microsoft.apimanagement/service/loggers.json` ARM template](nested/microsoft.apimanagement/service/loggers.json)
+
+## Troubleshooting
+
+- It is possible that the final step of deployment `app-service-webjob-msdeploy` reports a failed deployment with error such as `conflict` or `BadRequest` or `GatewayTimeout`. Despite these errors, it is possible that the deployment may have actually succeeded, so you may need to ignore this error. You may view detailed logs in your App Service/Activity log
```

Comment:
Why?  This sounds like a bug...

Reply:
Here are the repro steps:
1. Create App Service Plan (`serverfarm`)(not shared) - Completes OK
2. Create App Service: (`sites`, `sites/config`, `sites/hostNameBindings`) with depends on `serverfarm` - Completes OK
3 Run MSDeploy to install WebJob (zip file) (`sites/extensions`) with depends on the above 2. 
This Step 3. often fails in deployment ARM. The reason reported can either be `BadRequest` or `GatewayTimeout`.
4. Ignore failure and DO NOTHING.. just wait for 5-10 minutes. And the Webjob is deployed and successfully runs.

By looking at detailed logs for `App Service` one can view details on the `BadRequest` and `GatewayTimeout` events. Sometimes I have seen `"MessageTemplate": "A TXT record pointing from asuid.{0} to {1} was not found."` - which indicates that DNS propagation may not have completed. Eventually, all these errors succeed and the webjob is successfully deployed - 5-10 mins after ARM template failed.

it appears that the `dependsOn` does not take into account DNS propagation activities, which fails msdeploy.

Would you happen to know more about this or have a pointer to somewhere else? If there was a gap of 3-7 mins between step 2 and 3, the deployment would succeed.
I am updating this PR with changes shortly after I answer all questions below. Thx.